File: mapwidget/javascript/leaflet.js
// import ESM version of Leaflet
import * as L from "https://unpkg.com/leaflet@1.9.3/dist/leaflet-src.esm.js";
export function render(view) {
    // Header
    let center = view.model.get("center");
    let zoom = view.model.get("zoom");
    let width = view.model.get("width");
    let height = view.model.get("height");

    const container = document.createElement("div");
    container.style.width = width;
    container.style.height = height;

    // Map content

    const map = L.map(container).setView(center, zoom);
    L.tileLayer("https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png", {
        attribution:
            'Map data &copy; <a href="https://www.openstreetmap.org/">OpenStreetMap</a> contributors',
        maxZoom: 18,
    }).addTo(map);

    map.on("click", function (e) {
        console.log("Clicked at:", e.latlng);
        view.model.set("clicked_latlng", [e.latlng.lat, e.latlng.lng]);
        view.model.save_changes();
    });

    map.on("moveend", function (e) {
        let center = map.getCenter();
        view.model.set("center", [center.lat, center.lng]);
        let bbox = map.getBounds();
        let bounds = [
            bbox._southWest.lng,
            bbox._southWest.lat,
            bbox._northEast.lng,
            bbox._northEast.lat,
        ];
        view.model.set("bounds", bounds);
        view.model.save_changes();
    });

    map.on("zoomend", function (e) {
        let center = map.getCenter();
        view.model.set("center", [center.lat, center.lng]);
        view.model.set("zoom", map.getZoom());
        let bbox = map.getBounds();
        let bounds = [
            bbox._southWest.lng,
            bbox._southWest.lat,
            bbox._northEast.lng,
            bbox._northEast.lat,
        ];
        view.model.set("bounds", bounds);
        view.model.save_changes();
    });

    view.model.on("msg:custom", (msg) => {
        switch (msg.type) {
            case "add_basemap":
                L.tileLayer(msg.url, msg).addTo(map);
                break;
            case "add_layer":
                L.tileLayer(msg.url, msg).addTo(map);
                break;
            default:
                console.err(`Unsupported message '${msg.type}'.`);
        }
    });

    // Footer
    view.el.appendChild(container);
}

File: mapwidget/leaflet.py
import os
import pathlib
import anywidget
import traitlets


class Map(anywidget.AnyWidget):
    """Map widget

    Args:
        anywidget (_type_): _description_
    """

    _cwd = os.path.dirname(os.path.abspath(__file__))
    _esm = pathlib.Path(os.path.join(_cwd, 'javascript', 'leaflet.js'))
    _css = pathlib.Path(os.path.join(_cwd, 'styles', 'leaflet.css'))
    center = traitlets.List([40, -100]).tag(sync=True, o=True)
    zoom = traitlets.Int(4).tag(sync=True, o=True)
    bounds = traitlets.List([0, 0, 0, 0]).tag(sync=True, o=True)
    width = traitlets.Unicode('100%').tag(sync=True, o=True)
    height = traitlets.Unicode('600px').tag(sync=True, o=True)
    clicked_latlng = traitlets.List([None, None]).tag(sync=True, o=True)

    def add_basemap(self, name, opacity=1.0, **kwargs):
        from .basemaps import get_xyz_dict

        xyz_tiles = get_xyz_dict()

        if name in xyz_tiles:
            url = xyz_tiles[name]['url']
            attribution = xyz_tiles[name]['attribution']
            max_zoom = xyz_tiles[name]['max_zoom']
            self.send(
                {
                    "type": "add_basemap",
                    "url": url,
                    "attribution": attribution,
                    "maxZoom": max_zoom,
                    "opacity": opacity,
                    "name": name,
                }
            )

        else:
            raise ValueError(
                f"Basemap {name} not found. It must be one of the following: {list(xyz_tiles.keys())}"
            )

    def add_layer(
        self, url, name='Layer', attribution='', max_zoom=24, opacity=1.0, **kwargs
    ):
        self.send(
            {
                "type": "add_layer",
                "url": url,
                "attribution": attribution,
                "maxZoom": max_zoom,
                "name": name,
                "opacity": opacity,
            }
        )
